File: distrib/climate_analysis/test_sst_anomaly.m
%{
test_sst_anomaly.m - Climate mode analysis of NOAA SST anomalies using SVD.

Description
-----------
Applies EOF/POD analysis to Sea Surface Temperature anomalies to extract
dominant climate modes:

  1. EOF1: Trend/mean residual pattern
  2. EOF2: ENSO (El Nino-Southern Oscillation) - tropical Pacific
  3. EOF3: Pacific Decadal Oscillation (PDO) or other modes

This is the classic application of randomized SVD to climate data,
as studied by Tropp and others.

Preprocessing:
  1. Compute monthly climatology (mean January, mean February, etc.)
  2. Compute anomalies (deviation from climatology)

Data: SST anomalies in C (seasonal cycle removed)

EOF (Empirical Orthogonal Functions) = PCA = POD = SVD on anomaly data

Prerequisites
-------------
Run download_sst_data.m first:
  download_sst_data

Requirements
------------
* MATLAB R2014b+ with NetCDF support (built-in)
* Statistics and Machine Learning Toolbox (for corr function)
* librla.m (automatically added to path)
%}

% Add parent directory to path for librla
addpath(fullfile(fileparts(mfilename('fullpath')), '..', 'matlab'));

% ========================================================================
% Load SST data
% ========================================================================

fprintf('%s\n', repmat('=', 1, 70));
fprintf('NOAA SST Climate Mode Analysis (EOF/SVD on Anomalies)\n');
fprintf('%s\n', repmat('=', 1, 70));

DATA_DIR = fullfile(fileparts(mfilename('fullpath')), 'data');

% Find all downloaded files
nc_files = dir(fullfile(DATA_DIR, 'ersst.v5.*.nc'));
nc_files = sort({nc_files.name});

if isempty(nc_files)
    error(['No data files found in ', DATA_DIR, '\n\n', ...
           'Please run download_sst_data.m first:\n', ...
           '  download_sst_data']);
end

fprintf('\nFound %d monthly SST files\n', length(nc_files));
fprintf('  First: %s\n', nc_files{1});
fprintf('  Last:  %s\n', nc_files{end});

% Load first file to get grid info
first_file = fullfile(DATA_DIR, nc_files{1});
lon = ncread(first_file, 'lon');
lat = ncread(first_file, 'lat');

n_lon = length(lon);
n_lat = length(lat);
n_time = length(nc_files);

fprintf('\nGrid dimensions:\n');
fprintf('  Longitude: %d points (%.1f to %.1f)\n', n_lon, lon(1), lon(end));
fprintf('  Latitude:  %d points (%.1f to %.1f)\n', n_lat, lat(1), lat(end));
fprintf('  Time:      %d months\n', n_time);

% Load all SST data
fprintf('\nLoading SST data...\n');
SST = zeros(n_lon, n_lat, n_time, 'single');
dates = NaT(n_time, 1);

for i = 1:n_time
    f = fullfile(DATA_DIR, nc_files{i});
    sst_raw = ncread(f, 'sst');
    SST(:, :, i) = single(sst_raw(:, :, 1, 1));  % lon x lat x lev x time

    % Extract date from filename (ersst.v5.YYYYMM.nc)
    tokens = regexp(nc_files{i}, 'ersst\.v5\.(\d{4})(\d{2})\.nc', 'tokens');
    if ~isempty(tokens)
        dates(i) = datetime(str2double(tokens{1}{1}), str2double(tokens{1}{2}), 1);
    end

    if mod(i, 100) == 0
        fprintf('  Loaded %d / %d\r', i, n_time);
    end
end
fprintf('  Loaded %d months              \n', n_time);

% Handle missing values (land = NaN in ERSST)
SST(SST < -900) = NaN;  % Missing value flag

% Center on Greenwich meridian: convert 0-360 to -180 to 180
fprintf('\nCentering on Greenwich meridian...\n');
shift_idx = find(lon >= 180, 1);
lon = [lon(shift_idx:end) - 360; lon(1:shift_idx-1)];
SST = cat(1, SST(shift_idx:end, :, :), SST(1:shift_idx-1, :, :));

fprintf('  Longitude: %.1f to %.1f\n', lon(1), lon(end));
valid_sst = SST(~isnan(SST));
fprintf('\nSST range: %.1f C to %.1f C\n', min(valid_sst), max(valid_sst));

% ========================================================================
% Compute climatology and anomalies
% ========================================================================

fprintf('\n%s\n', repmat('=', 1, 70));
fprintf('Computing climatology and anomalies...\n');
fprintf('%s\n', repmat('=', 1, 70));

% Monthly climatology (average January, average February, etc.)
climatology = zeros(n_lon, n_lat, 12, 'single');
for m = 1:12
    month_idx = find(month(dates) == m);
    climatology(:, :, m) = mean(SST(:, :, month_idx), 3);
end

% Compute anomalies (deviation from climatology)
SST_anom = zeros(size(SST), 'single');
for i = 1:n_time
    SST_anom(:, :, i) = SST(:, :, i) - climatology(:, :, month(dates(i)));
end

valid_anom = SST_anom(~isnan(SST_anom));
fprintf('  Anomaly range: %.2f C to %.2f C\n', min(valid_anom), max(valid_anom));

% ========================================================================
% Create ocean mask and reshape for SVD
% ========================================================================

fprintf('\n%s\n', repmat('=', 1, 70));
fprintf('Preparing data matrix...\n');
fprintf('%s\n', repmat('=', 1, 70));

% Ocean mask (where we have valid data for all times)
ocean_mask = ~any(isnan(SST_anom), 3);
n_ocean = sum(ocean_mask(:));

fprintf('\n  Ocean points: %d / %d (%.1f%%)\n', n_ocean, n_lon * n_lat, ...
        100*n_ocean/(n_lon*n_lat));

% Reshape to 2D matrix: ocean_points x time
SST_matrix = zeros(n_ocean, n_time, 'single');
ocean_idx = find(ocean_mask);

for t = 1:n_time
    anom_t = SST_anom(:, :, t);
    SST_matrix(:, t) = anom_t(ocean_idx);
end

fprintf('  Data matrix: %d x %d\n', n_ocean, n_time);

% ========================================================================
% EOF Analysis via SVD
% ========================================================================

fprintf('\n%s\n', repmat('=', 1, 70));
fprintf('EOF Analysis (SVD)\n');
fprintf('%s\n', repmat('=', 1, 70));

% Number of modes to compute
n_modes = 30;

fprintf('\nComputing %d-mode SVD using svd_sketch...\n', n_modes);
tic;
[U, s, V] = librla.svd_sketch(double(SST_matrix), n_modes, 'power_iter', 3, 'extra_samples', 20);
fprintf('  Elapsed time: %.2fs\n', toc);

% Also compute reference SVD
fprintf('\nComputing reference SVD...\n');
tic;
[U_full, s_full, V_full] = svd(double(SST_matrix), 'econ');
s_full = diag(s_full);
fprintf('  Elapsed time: %.2fs\n', toc);

% Variance explained
total_var = sum(s_full.^2);
var_explained = s_full.^2 / total_var;
cumulative_var = cumsum(var_explained);

fprintf('\nEOF variance explained:\n');
fprintf('  Mode    Variance %%   Cumulative %%\n');
fprintf('  %s\n', repmat('-', 1, 40));
for i = 1:min(10, n_modes)
    fprintf('  EOF%2d   %6.2f%%      %6.2f%%\n', i, 100*var_explained(i), 100*cumulative_var(i));
end

n90 = find(cumulative_var >= 0.90, 1);
n95 = find(cumulative_var >= 0.95, 1);
fprintf('\n  Modes for 90%%: %d, 95%%: %d\n', n90, n95);

% ========================================================================
% Reshape EOFs back to spatial maps
% ========================================================================

% Compute Nino 3.4 index for ENSO validation (not for sign fixing)
% Nino 3.4 region: 5N-5S, 170W-120W (-170 to -120 in -180/180 system)
nino34_lon_idx = find(lon >= -170 & lon <= -120);
nino34_lat_idx = find(lat >= -5 & lat <= 5);

nino34_index = zeros(n_time, 1);
for t = 1:n_time
    region = SST_anom(nino34_lon_idx, nino34_lat_idx, t);
    nino34_index(t) = mean(region(~isnan(region)));
end

% Fix signs based on spatial pattern mean
% Convention: spatial pattern should have positive mean over ocean
% This ensures positive PC = positive anomaly contribution

% Deterministic SVD
U_det = U_full(:, 1:n_modes);
V_det = V_full(:, 1:n_modes);
for i = 1:n_modes
    if mean(U_det(:, i)) < 0
        U_det(:, i) = -U_det(:, i);
        V_det(:, i) = -V_det(:, i);
    end
end

% Randomized SVD
U_rand = U;
V_rand = V;
for i = 1:n_modes
    if mean(U_rand(:, i)) < 0
        U_rand(:, i) = -U_rand(:, i);
        V_rand(:, i) = -V_rand(:, i);
    end
end

% Spatial: U (non-dimensional pattern)
% Temporal: projection of data onto U, normalized by sqrt(n_ocean) for C units
EOF_maps_det = cell(n_modes, 1);
PC_det = cell(n_modes, 1);
EOF_maps_rand = cell(n_modes, 1);
PC_rand = cell(n_modes, 1);

for i = 1:n_modes
    % Reshape to map
    field_det = NaN(n_lon, n_lat, 'single');
    field_det(ocean_idx) = U_det(:, i);
    EOF_maps_det{i} = field_det;
    PC_det{i} = (U_det(:, i)' * double(SST_matrix))' / sqrt(n_ocean);

    field_rand = NaN(n_lon, n_lat, 'single');
    field_rand(ocean_idx) = U_rand(:, i);
    EOF_maps_rand{i} = field_rand;
    PC_rand{i} = (U_rand(:, i)' * double(SST_matrix))' / sqrt(n_ocean);
end

% ========================================================================
% Visualize Deterministic SVD results
% ========================================================================

fprintf('\n%s\n', repmat('=', 1, 70));
fprintf('Visualizing EOF patterns...\n');
fprintf('%s\n', repmat('=', 1, 70));

% Convert dates to decimal years for plotting
years = year(dates) + (month(dates) - 1) / 12;

% Mode labels
mode_names = {'EOF1 - Trend', 'EOF2 - ENSO', 'EOF3 - PDO', 'EOF4', 'EOF5'};

figure('Name', 'Deterministic SVD (Anomalies)', 'Position', [50, 50, 1400, 1000]);
for i = 1:5
    % Left: Spatial EOF pattern
    subplot(5, 2, 2*i-1);
    clim_val = max(abs(EOF_maps_det{i}(:)), [], 'omitnan') * 0.8;
    pcolor(lon, lat, EOF_maps_det{i}');
    shading flat;
    colormap(gca, redblue());
    caxis([-clim_val, clim_val]);
    xlabel('Longitude');
    ylabel('Latitude');
    title(sprintf('%s (%.1f%%)', mode_names{i}, 100*var_explained(i)));
    axis equal tight;

    % Right: Temporal PC coefficient
    subplot(5, 2, 2*i);
    plot(years, PC_det{i}, 'LineWidth', 0.8, 'Color', [0.27, 0.51, 0.71]);
    xlabel('Year');
    ylabel(sprintf('PC%d (C)', i));
    yline(0, '--', 'Color', [0.5, 0.5, 0.5]);
    xlim([min(years), max(years)]);
end
sgtitle('Deterministic SVD (Anomalies)', 'FontWeight', 'bold');

% ========================================================================
% Visualize Randomized SVD results
% ========================================================================

var_explained_rand = s.^2 / total_var;

figure('Name', 'Randomized SVD (Anomalies)', 'Position', [100, 100, 1400, 1000]);
for i = 1:5
    % Left: Spatial EOF pattern
    subplot(5, 2, 2*i-1);
    clim_val = max(abs(EOF_maps_rand{i}(:)), [], 'omitnan') * 0.8;
    pcolor(lon, lat, EOF_maps_rand{i}');
    shading flat;
    colormap(gca, redblue());
    caxis([-clim_val, clim_val]);
    xlabel('Longitude');
    ylabel('Latitude');
    title(sprintf('%s (%.1f%%)', mode_names{i}, 100*var_explained_rand(i)));
    axis equal tight;

    % Right: Temporal PC coefficient
    subplot(5, 2, 2*i);
    plot(years, PC_rand{i}, 'LineWidth', 0.8, 'Color', [0.27, 0.51, 0.71]);
    xlabel('Year');
    ylabel(sprintf('PC%d (C)', i));
    yline(0, '--', 'Color', [0.5, 0.5, 0.5]);
    xlim([min(years), max(years)]);
end
sgtitle('Randomized SVD (svd\_sketch)', 'FontWeight', 'bold');

% ========================================================================
% Singular value spectrum and variance comparison
% ========================================================================

figure('Name', 'Singular Values', 'Position', [150, 150, 600, 400]);

% Singular value spectrum
semilogy(1:min(50, length(s_full)), s_full(1:min(50, length(s_full))), 'o', ...
         'MarkerSize', 4, 'DisplayName', 'Deterministic');
hold on;
semilogy(1:n_modes, s, 'x', 'MarkerSize', 8, 'Color', 'r', 'DisplayName', 'Randomized');
hold off;
xlabel('Mode');
ylabel('\sigma_i');
title('Singular Value Spectrum');
legend('Location', 'northeast');

% ========================================================================
% ENSO region analysis
% ========================================================================

fprintf('\n%s\n', repmat('=', 1, 70));
fprintf('ENSO Analysis\n');
fprintf('%s\n', repmat('=', 1, 70));

% Correlation between PC2 (ENSO mode) and Nino 3.4
corr_pc2_nino = corr(PC_det{2}, nino34_index);
fprintf('\nCorrelation between PC2 and Nino 3.4 index: %.3f\n', corr_pc2_nino);

% Plot comparison
figure('Name', 'ENSO Comparison', 'Position', [200, 200, 900, 400]);
plot(years, PC_det{2} / std(PC_det{2}), 'LineWidth', 1.5, 'Color', 'b', 'DisplayName', 'PC2 (det)');
hold on;
plot(years, nino34_index / std(nino34_index), '--', 'LineWidth', 1.5, 'Color', 'r', 'DisplayName', 'Nino 3.4');
yline(0, ':', 'Color', [0.5, 0.5, 0.5]);
hold off;
xlabel('Year');
ylabel('Index (normalized)');
title(sprintf('PC2 (ENSO) vs Nino 3.4 Index (r = %.2f)', corr_pc2_nino));
legend('Location', 'northwest');
xlim([min(years), max(years)]);

% ========================================================================
% Summary
% ========================================================================

fprintf('\n%s\n', repmat('=', 1, 70));
fprintf('Summary\n');
fprintf('%s\n', repmat('=', 1, 70));
fprintf('  Data: NOAA ERSST v5 (anomalies)\n');
fprintf('  Grid: %d x %d (2 deg resolution)\n', n_lon, n_lat);
fprintf('  Time: %s to %s (%d months)\n', datestr(dates(1), 'yyyy-mm'), ...
        datestr(dates(end), 'yyyy-mm'), n_time);
fprintf('  Ocean points: %d\n', n_ocean);
fprintf('\n');
fprintf('  EOF Analysis:\n');
fprintf('    EOF1 variance: %.1f%% (trend/mean)\n', 100*var_explained(1));
fprintf('    EOF2 variance: %.1f%% (ENSO)\n', 100*var_explained(2));
fprintf('    Modes for 90%% variance: %d\n', n90);
fprintf('\n');
fprintf('  Randomized SVD accuracy:\n');
rel_err = norm(s - s_full(1:n_modes)) / norm(s_full(1:n_modes));
fprintf('    Singular value error: %.3f%%\n', 100*rel_err);
fprintf('\n');
fprintf('  ENSO validation (PC2 - Nino 3.4 correlation):\n');
corr_pc2_rand = corr(PC_rand{2}, nino34_index);
fprintf('    Deterministic SVD: %.3f\n', corr_pc2_nino);
fprintf('    Randomized SVD:    %.3f\n', abs(corr_pc2_rand));
fprintf('%s\n', repmat('=', 1, 70));

% ========================================================================
% Helper function: Red-Blue colormap
% ========================================================================
function cmap = redblue(m)
    if nargin < 1
        m = 256;
    end
    % Create a red-white-blue colormap
    r = [linspace(0, 1, m/2), ones(1, m/2)]';
    g = [linspace(0, 1, m/2), linspace(1, 0, m/2)]';
    b = [ones(1, m/2), linspace(1, 0, m/2)]';
    cmap = [r, g, b];
end

File: distrib/climate_analysis/test_sst_anomaly.py
"""
test_sst_anomaly.py - Climate mode analysis of NOAA SST anomalies using SVD.

Description
-----------
Applies EOF/POD analysis to Sea Surface Temperature anomalies to extract
dominant climate modes:

  1. EOF1: Trend/mean residual pattern
  2. EOF2: ENSO (El Niño-Southern Oscillation) - tropical Pacific
  3. EOF3: Other modes

This is the classic application of randomized SVD to climate data,
as studied by Tropp and others.

Preprocessing:
  1. Compute monthly climatology (mean January, mean February, etc.)
  2. Compute anomalies (deviation from climatology)

Data: SST anomalies in °C (seasonal cycle removed)

EOF (Empirical Orthogonal Functions) = PCA = POD = SVD on anomaly data

Prerequisites
-------------
Run download_sst_data.py first:
  python download_sst_data.py

Requirements
------------
* librla.py in the path
* pip install numpy scipy netCDF4 matplotlib
"""

import sys
import os
import re
import time
from pathlib import Path
from datetime import datetime

import numpy as np
from numpy.linalg import norm
from scipy.stats import pearsonr
import netCDF4 as nc
import matplotlib.pyplot as plt

# Add parent directory to path for librla
sys.path.insert(0, str(Path(__file__).parent.parent / "python"))
import librla

# ========================================================================
# Load SST data
# ========================================================================

print("=" * 70)
print("NOAA SST Climate Mode Analysis (EOF/SVD on Anomalies)")
print("=" * 70)

DATA_DIR = Path(__file__).parent / "data"

# Find all downloaded files
nc_files = sorted(DATA_DIR.glob("ersst.v5.*.nc"))

if not nc_files:
    raise FileNotFoundError(f"""
    No data files found in {DATA_DIR}

    Please run download_sst_data.py first:
      python download_sst_data.py
    """)

print(f"\nFound {len(nc_files)} monthly SST files")
print(f"  First: {nc_files[0].name}")
print(f"  Last:  {nc_files[-1].name}")

# Load first file to get grid info
ds = nc.Dataset(nc_files[0])
lon = ds.variables['lon'][:]
lat = ds.variables['lat'][:]
ds.close()

n_lon = len(lon)
n_lat = len(lat)
n_time = len(nc_files)

print(f"\nGrid dimensions:")
print(f"  Longitude: {n_lon} points ({lon[0]}° to {lon[-1]}°)")
print(f"  Latitude:  {n_lat} points ({lat[0]}° to {lat[-1]}°)")
print(f"  Time:      {n_time} months")

# Load all SST data
print("\nLoading SST data...")
SST = np.zeros((n_lon, n_lat, n_time), dtype=np.float32)
dates = []

for i, f in enumerate(nc_files):
    ds = nc.Dataset(f)
    sst_raw = ds.variables['sst'][0, 0, :, :]  # time × lev × lat × lon -> lat × lon
    # NetCDF returns masked array - convert to NaN
    SST[:, :, i] = np.ma.filled(sst_raw.T, np.nan)  # Transpose to lon × lat

    # Extract date from filename (ersst.v5.YYYYMM.nc)
    m = re.match(r"ersst\.v5\.(\d{4})(\d{2})\.nc", f.name)
    if m:
        dates.append(datetime(int(m.group(1)), int(m.group(2)), 1))
    ds.close()

    if (i + 1) % 100 == 0:
        print(f"  Loaded {i + 1} / {n_time}\r", end="", flush=True)

print(f"  Loaded {n_time} months              ")

# Handle missing values (land = NaN in ERSST)
SST[SST < -900] = np.nan  # Missing value flag

# Center on Greenwich meridian: convert 0-360 to -180 to 180
print("\nCentering on Greenwich meridian...")
shift_idx = np.argmax(lon >= 180)
lon = np.concatenate([lon[shift_idx:] - 360, lon[:shift_idx]])
SST = np.concatenate([SST[shift_idx:, :, :], SST[:shift_idx, :, :]], axis=0)

print(f"  Longitude: {lon[0]}° to {lon[-1]}°")
valid_sst = SST[~np.isnan(SST)]
print(f"\nSST range: {np.min(valid_sst):.1f}°C to {np.max(valid_sst):.1f}°C")

# ========================================================================
# Compute climatology and anomalies
# ========================================================================

print("\n" + "=" * 70)
print("Computing climatology and anomalies...")
print("=" * 70)

# Monthly climatology (average January, average February, etc.)
climatology = np.zeros((n_lon, n_lat, 12), dtype=np.float32)
for month in range(12):
    month_idx = [i for i, d in enumerate(dates) if d.month == month + 1]
    climatology[:, :, month] = np.mean(SST[:, :, month_idx], axis=2)

# Compute anomalies (deviation from climatology)
SST_anom = np.zeros_like(SST)
for i, d in enumerate(dates):
    SST_anom[:, :, i] = SST[:, :, i] - climatology[:, :, d.month - 1]

valid_anom = SST_anom[~np.isnan(SST_anom)]
print(f"  Anomaly range: {np.min(valid_anom):.2f}°C to {np.max(valid_anom):.2f}°C")

# ========================================================================
# Create ocean mask and reshape for SVD
# ========================================================================

print("\n" + "=" * 70)
print("Preparing data matrix...")
print("=" * 70)

# Ocean mask (where we have valid data for all times)
ocean_mask = ~np.any(np.isnan(SST_anom), axis=2)
n_ocean = np.sum(ocean_mask)

print(f"\n  Ocean points: {n_ocean} / {n_lon * n_lat} ({100*n_ocean/(n_lon*n_lat):.1f}%)")

# Reshape to 2D matrix: ocean_points × time
SST_matrix = np.zeros((n_ocean, n_time), dtype=np.float32)

for t in range(n_time):
    SST_matrix[:, t] = SST_anom[:, :, t][ocean_mask]

print(f"  Data matrix: {n_ocean} × {n_time}")

# ========================================================================
# EOF Analysis via SVD
# ========================================================================

print("\n" + "=" * 70)
print("EOF Analysis (SVD)")
print("=" * 70)

# Number of modes to compute
n_modes = 30

print(f"\nComputing {n_modes}-mode SVD using svd_sketch...")
t0 = time.time()
U, s, Vh = librla.svd_sketch(SST_matrix, n_modes, power_iter=3, extra_samples=20)
V = Vh.T
print(f"  Elapsed time: {time.time() - t0:.2f}s")

# Also compute reference SVD
print("\nComputing reference SVD...")
t0 = time.time()
U_full, s_full, Vh_full = np.linalg.svd(SST_matrix, full_matrices=False)
V_full = Vh_full.T
print(f"  Elapsed time: {time.time() - t0:.2f}s")

# Variance explained
total_var = np.sum(s_full**2)
var_explained = s_full**2 / total_var
cumulative_var = np.cumsum(var_explained)

print("\nEOF variance explained:")
print("  Mode    Variance %   Cumulative %")
print("  " + "-" * 40)
for i in range(min(10, n_modes)):
    print(f"  EOF{i+1:2d}   {100*var_explained[i]:6.2f}%      {100*cumulative_var[i]:6.2f}%")

n90 = np.argmax(cumulative_var >= 0.90) + 1
n95 = np.argmax(cumulative_var >= 0.95) + 1
print(f"\n  Modes for 90%: {n90}, 95%: {n95}")

# ========================================================================
# Reshape EOFs back to spatial maps
# ========================================================================

def reshape_to_map(vec, mask):
    """Reshape ocean vector back to lon×lat map with NaN for land."""
    field = np.full((len(lon), len(lat)), np.nan, dtype=np.float32)
    field[mask] = vec
    return field


# Compute Niño 3.4 index for ENSO validation (not for sign fixing)
# Niño 3.4 region: 5°N-5°S, 170°W-120°W (-170° to -120° in -180/180 system)
nino34_lon_idx = np.where((lon >= -170) & (lon <= -120))[0]
nino34_lat_idx = np.where((lat >= -5) & (lat <= 5))[0]

nino34_index = np.zeros(n_time)
for t in range(n_time):
    region = SST_anom[np.ix_(nino34_lon_idx, nino34_lat_idx, [t])][:, :, 0]
    nino34_index[t] = np.nanmean(region)


def fix_svd_signs(U_modes, V_modes):
    """Fix signs based on spatial pattern mean.

    Convention: spatial pattern should have positive mean over ocean.
    This ensures positive PC = positive anomaly contribution.
    """
    U_modes = U_modes.copy()
    V_modes = V_modes.copy()
    for i in range(U_modes.shape[1]):
        if np.mean(U_modes[:, i]) < 0:
            U_modes[:, i] *= -1
            V_modes[:, i] *= -1
    return U_modes, V_modes


# Deterministic SVD
U_det, V_det = fix_svd_signs(U_full[:, :n_modes], V_full[:, :n_modes])

# Randomized SVD
U_rand, V_rand = fix_svd_signs(U, V)

# Spatial: U (non-dimensional pattern)
# Temporal: projection of data onto U, normalized by √n_ocean for °C units
EOF_maps_det = [reshape_to_map(U_det[:, i], ocean_mask) for i in range(n_modes)]
PC_det = [(U_det[:, i] @ SST_matrix) / np.sqrt(n_ocean) for i in range(n_modes)]

EOF_maps_rand = [reshape_to_map(U_rand[:, i], ocean_mask) for i in range(n_modes)]
PC_rand = [(U_rand[:, i] @ SST_matrix) / np.sqrt(n_ocean) for i in range(n_modes)]


# ========================================================================
# Visualize Deterministic SVD results
# ========================================================================

print("\n" + "=" * 70)
print("Visualizing EOF patterns...")
print("=" * 70)

# Convert dates to decimal years for plotting
years = np.array([d.year + (d.month - 1) / 12 for d in dates])

# Mode labels
mode_names = ["EOF1 - Trend", "EOF2 - ENSO", "EOF3", "EOF4", "EOF5"]

fig1, axes1 = plt.subplots(5, 2, figsize=(12, 14))
fig1.suptitle("Deterministic SVD (Anomalies)", fontsize=14, fontweight='bold')

for i in range(5):
    # Left: Spatial EOF pattern
    ax_spatial = axes1[i, 0]
    clim = np.nanmax(np.abs(EOF_maps_det[i])) * 0.8
    im = ax_spatial.pcolormesh(lon, lat, EOF_maps_det[i].T, cmap='RdBu_r',
                                vmin=-clim, vmax=clim, shading='auto')
    ax_spatial.set_xlabel("Longitude")
    ax_spatial.set_ylabel("Latitude")
    ax_spatial.set_title(f"{mode_names[i]} ({100*var_explained[i]:.1f}%)")

    # Right: Temporal PC coefficient
    ax_temporal = axes1[i, 1]
    ax_temporal.plot(years, PC_det[i], linewidth=0.8, color='steelblue')
    ax_temporal.set_xlabel("Year")
    ax_temporal.set_ylabel(f"PC{i+1} (°C)")
    ax_temporal.axhline(0, color='gray', linestyle='--', linewidth=0.5)

fig1.subplots_adjust(top=0.95, bottom=0.05, hspace=0.35, wspace=0.25)

# ========================================================================
# Visualize Randomized SVD results
# ========================================================================

var_explained_rand = s**2 / total_var

fig2, axes2 = plt.subplots(5, 2, figsize=(12, 14))
fig2.suptitle("Randomized SVD (Anomalies)", fontsize=14, fontweight='bold')

for i in range(5):
    # Left: Spatial EOF pattern
    ax_spatial = axes2[i, 0]
    clim = np.nanmax(np.abs(EOF_maps_rand[i])) * 0.8
    im = ax_spatial.pcolormesh(lon, lat, EOF_maps_rand[i].T, cmap='RdBu_r',
                                vmin=-clim, vmax=clim, shading='auto')
    ax_spatial.set_xlabel("Longitude")
    ax_spatial.set_ylabel("Latitude")
    ax_spatial.set_title(f"{mode_names[i]} ({100*var_explained_rand[i]:.1f}%)")

    # Right: Temporal PC coefficient
    ax_temporal = axes2[i, 1]
    ax_temporal.plot(years, PC_rand[i], linewidth=0.8, color='steelblue')
    ax_temporal.set_xlabel("Year")
    ax_temporal.set_ylabel(f"PC{i+1} (°C)")
    ax_temporal.axhline(0, color='gray', linestyle='--', linewidth=0.5)

fig2.subplots_adjust(top=0.95, bottom=0.05, hspace=0.35, wspace=0.25)

# ========================================================================
# Singular value spectrum and variance comparison
# ========================================================================

fig3, ax3 = plt.subplots(figsize=(8, 4))

# Singular value spectrum
n_plot = min(50, len(s_full))
ax3.semilogy(range(1, n_plot + 1), s_full[:n_plot],
             'o', markersize=4, label="Deterministic")
ax3.semilogy(range(1, n_modes + 1), s, 'x', markersize=8, color='red',
             label="Randomized")
ax3.set_xlabel("Mode")
ax3.set_ylabel("σᵢ")
ax3.set_title("Singular Value Spectrum")
ax3.legend()

fig3.tight_layout()

# ========================================================================
# ENSO region analysis
# ========================================================================

print("\n" + "=" * 70)
print("ENSO Analysis")
print("=" * 70)

# Niño 3.4 index was computed earlier for sign determination
# Correlation between PC2 (ENSO mode) and Niño 3.4
corr_pc2_nino = pearsonr(PC_det[1], nino34_index)[0]
print(f"\nCorrelation between PC2 and Niño 3.4 index: {corr_pc2_nino:.3f}")

# Plot comparison
fig4, ax4 = plt.subplots(figsize=(10, 3))
ax4.plot(years, PC_det[1] / np.std(PC_det[1]), linewidth=1.5, label="PC2 (det)", color='blue')
ax4.plot(years, nino34_index / np.std(nino34_index), linewidth=1.5,
         label="Niño 3.4", color='red', linestyle='--')
ax4.axhline(0, color='gray', linestyle=':', linewidth=0.5)
ax4.set_xlabel("Year")
ax4.set_ylabel("Index (normalized)")
ax4.set_title(f"PC2 (ENSO) vs Niño 3.4 Index (r = {corr_pc2_nino:.2f})")
ax4.legend()

fig4.tight_layout()

# ========================================================================
# Summary
# ========================================================================

print("\n" + "=" * 70)
print("Summary")
print("=" * 70)
print(f"  Data: NOAA ERSST v5 (anomalies)")
print(f"  Grid: {n_lon} × {n_lat} (2° resolution)")
print(f"  Time: {dates[0].strftime('%Y-%m')} to {dates[-1].strftime('%Y-%m')} ({n_time} months)")
print(f"  Ocean points: {n_ocean}")
print()
print("  EOF Analysis:")
print(f"    EOF1 variance: {100*var_explained[0]:.1f}% (trend/mean)")
print(f"    EOF2 variance: {100*var_explained[1]:.1f}% (ENSO)")
print(f"    Modes for 90% variance: {n90}")
print()
print("  Randomized SVD accuracy:")
rel_err = norm(s - s_full[:n_modes]) / norm(s_full[:n_modes])
print(f"    Singular value error: {100*rel_err:.3f}%")
print()
print("  ENSO validation (PC2 - Niño 3.4 correlation):")
corr_pc2_rand = pearsonr(PC_rand[1], nino34_index)[0]
print(f"    Deterministic SVD: {corr_pc2_nino:.3f}")
print(f"    Randomized SVD:    {abs(corr_pc2_rand):.3f}")
print("=" * 70)

fig1.savefig("sst_anomaly.png", dpi=150)
print("\nFigure saved to sst_anomaly.png")
fig2.savefig("sst_anomaly_sketch.png", dpi=150)
print("\nFigure saved to sst_anomaly_sketch.png")
plt.show(block=False)
